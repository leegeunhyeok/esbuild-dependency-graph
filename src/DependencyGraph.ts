import type { Metafile } from 'esbuild';
import { isExternal } from './helpers';
import type {
  EsbuildModule,
  ExternalModule,
  InternalModule,
  Module,
  ModuleId,
} from './types';

type ModuleDependencyGraph = Record<ModuleId, Module | undefined>;

export class DependencyGraph {
  private dependencyGraph: ModuleDependencyGraph = {};
  private INTERNAL__moduleIds: Record<string, number> = {};
  private INTERNAL__moduleId = 0;

  constructor(
    private metafile: Metafile,
    entryPoint: string,
    initialModuleIdMap?: Record<string, number>,
  ) {
    // Entry point module id is always `0`.
    this.INTERNAL__moduleIds[entryPoint] = this.INTERNAL__moduleId++;
    if (initialModuleIdMap) {
      this.INTERNAL__moduleIds = initialModuleIdMap;
      this.INTERNAL__moduleId =
        Math.max(...Object.values(initialModuleIdMap)) + 1;
    }
    this.generateDependencyGraph();
  }

  /**
   * Generate unique id for module.
   *
   * If already has id for path return cached id
   * else generate new one.
   */
  private generateUniqueModuleId(path: string): number {
    return (
      this.INTERNAL__moduleIds[path] ??
      (() => (this.INTERNAL__moduleIds[path] = this.INTERNAL__moduleId++))()
    );
  }

  /**
   * Create `ExternalModule` data.
   */
  private getExternalModule(moduleId: ModuleId, path: string): ExternalModule {
    return Object.defineProperties(
      {},
      {
        id: { value: moduleId },
        path: { value: path },
        __external: {
          enumerable: false,
          value: true,
        },
      },
    ) as ExternalModule;
  }

  /**
   * Enhance esbuild's input module to `InternalModule`.
   */
  private toInternalModule(
    esbuildModule: EsbuildModule,
    moduleId: ModuleId,
    path: string,
  ): InternalModule {
    return Object.defineProperties(esbuildModule, {
      id: { value: moduleId },
      path: { value: path },
      dependencies: { value: new Set() },
      inverseDependencies: { value: new Set() },
    }) as InternalModule;
  }

  /**
   * Get module by actual path in metafile.
   */
  private addNode(modulePath: string, external = false): Module {
    let id: ModuleId | undefined;

    if (
      (id = this.INTERNAL__moduleIds[modulePath]) &&
      id in this.dependencyGraph
    ) {
      return this.dependencyGraph[id]!;
    }

    if (external) {
      id = this.generateUniqueModuleId(modulePath);
      return (this.dependencyGraph[id] = this.getExternalModule(
        id,
        modulePath,
      ));
    } else if (this.metafile.inputs[modulePath]) {
      id = this.generateUniqueModuleId(modulePath);
      return (this.dependencyGraph[id] = this.toInternalModule(
        this.metafile.inputs[modulePath],
        id,
        modulePath,
      ));
    }

    throw new Error(`unable get module: '${modulePath}'`);
  }

  /**
   * Traverse modules for get dependencies.
   */
  private generateDependencyGraph(): void {
    for (const modulePath in this.metafile.inputs) {
      const currentModule = this.addNode(modulePath);

      if (isExternal(currentModule)) {
        continue;
      }

      for (const importModule of currentModule.imports) {
        const importedModule = this.addNode(
          importModule.path,
          importModule.external,
        );

        currentModule.dependencies.add(importedModule.id);

        if (!isExternal(importedModule)) {
          importedModule.inverseDependencies.add(currentModule.id);
        }
      }
    }
  }

  /**
   * Traverse modules for get invert dependencies.
   */
  private traverseInverseModules(moduleId: ModuleId): ModuleId[] {
    const queue: ModuleId[] = [moduleId];
    const visited: Record<ModuleId, boolean> = { [moduleId]: true };
    const inverseModuleIds: ModuleId[] = [];

    while (queue.length) {
      const currentModuleId = queue.shift()!;
      const module = this.dependencyGraph[currentModuleId];
      inverseModuleIds.push(currentModuleId);

      if (module && isExternal(module)) continue;

      module?.inverseDependencies.forEach((inverseModuleId) => {
        if (visited[inverseModuleId]) return;
        visited[inverseModuleId] = true;
        queue.push(inverseModuleId);
      });
    }

    return inverseModuleIds;
  }

  /**
   * Get module id by actual module path.
   *
   * ```ts
   * // `Metafile` type in esbuild
   * interface Metafile {
   *   inputs: {
   *     [path: string]: { // Can be used as `modulePath`
   *       imports: {
   *         path: string // Can be used as `modulePath`
   *         ...
   *       }[]
   *       ...
   *     }
   *   },
   *   outputs: {
   *    ...
   *   }
   * }
   * ```
   */
  getModuleId(modulePath: string): number {
    const id = this.INTERNAL__moduleIds[modulePath];
    if (typeof id === 'number') return id;
    throw new Error(`unable get id: '${modulePath}'`);
  }

  /**
   * Get module by id.
   */
  getModule(moduleId: ModuleId): Module {
    const module = this.dependencyGraph[moduleId];
    if (module) return module;
    throw new Error(`unable get module: '${moduleId}'`);
  }

  /**
   * Get dependencies of specified module.
   */
  dependenciesOf(moduleId: ModuleId): ModuleId[] {
    const module = this.getModule(moduleId);
    return isExternal(module) ? [] : Array.from(module.dependencies);
  }

  /**
   * Get inverse dependencies of specified module.
   */
  inverseDependenciesOf(moduleId: ModuleId): ModuleId[] {
    return this.traverseInverseModules(moduleId);
  }
}
