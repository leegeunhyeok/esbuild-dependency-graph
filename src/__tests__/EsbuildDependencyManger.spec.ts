import { readFile } from 'node:fs/promises';
import { join } from 'node:path';
import { describe } from '@jest/globals';
import type { Metafile } from 'esbuild';
import { faker } from '@faker-js/faker';
import { EsbuildDependencyManager } from '../DependencyGraph';
import type { ModuleId, ModuleMap, ModuleDependencyGraph } from '../types';

const ENTRY_POINT = 'index.js';
const TEST_MODULE = 'src/screens/MainScreen.tsx';

beforeAll(() => {
  // noop
  console.warn = () => undefined;
});

describe('EsbuildDependencyManager', () => {
  let metafile: Metafile;
  let dependencyManager: EsbuildDependencyManager;
  let moduleDependencyGraph: ModuleDependencyGraph;

  beforeAll(async () => {
    metafile = JSON.parse(
      await readFile(join(__dirname, './fixtures/metafile.json'), 'utf-8'),
    ) as Metafile;
    dependencyManager = new EsbuildDependencyManager(metafile, ENTRY_POINT);
    moduleDependencyGraph = dependencyManager.getDependencyGraph();
  });

  describe('getModuleId', () => {
    describe('when trying to get a module id that does not exist', () => {
      it('should returns `null`', () => {
        expect(dependencyManager.getModuleId('-')).toBeNull();
      });
    });

    describe('when trying to get a module id that exist', () => {
      let modulePath: string;
      let moduleId: ModuleId;

      beforeEach(() => {
        modulePath = faker.helpers.arrayElement(Object.keys(metafile.inputs));
        moduleId = dependencyManager.getModuleId(modulePath)!;
      });

      it('should returns expected module', () => {
        expect(moduleId).not.toBeNull();
        expect(typeof moduleId === 'number').toEqual(true);
      });

      describe('when trying to get module by `moduleId`', () => {
        it('should returns expected module', () => {
          expect(dependencyManager.getModuleById(moduleId)?.path).toEqual(
            modulePath,
          );
        });
      });
    });
  });

  describe('getModule', () => {
    describe('when trying to get a module that does not exist', () => {
      it('should returns `null`', () => {
        expect(dependencyManager.getModule('-')).toBeNull();
      });
    });

    describe('when trying to get a module that exist', () => {
      let modulePath: string;

      beforeEach(() => {
        modulePath = faker.helpers.arrayElement(Object.keys(metafile.inputs));
      });

      it('should returns expected module', () => {
        const module = dependencyManager.getModule(modulePath);
        expect(module).not.toBeNull();
        expect(module?.path).toEqual(modulePath);
      });
    });
  });

  describe('getModuleMap', () => {
    let modulePath: string;
    let moduleId: ModuleId;
    let moduleMap: ModuleMap;

    beforeEach(() => {
      modulePath = faker.helpers.arrayElement(Object.keys(metafile.inputs));
      moduleId = dependencyManager.getModuleId(modulePath)!;
      moduleMap = dependencyManager.getModuleMap();
    });

    it('should returns `ModuleMap` object', () => {
      expect(typeof moduleMap === 'object').toEqual(true);
    });

    describe('when access by module id to `ModuleMap`', () => {
      it('should be same as the expected module', () => {
        expect(moduleMap[moduleId].path).toEqual(modulePath);
      });
    });
  });

  describe('getDependencyGraph', () => {
    let modulePath: string;
    let moduleId: ModuleId;

    beforeEach(() => {
      modulePath = faker.helpers.arrayElement(Object.keys(metafile.inputs));
      moduleId = dependencyManager.getModuleId(modulePath)!;
    });

    it('should returns `ModuleDependencyGraph` object', () => {
      expect(typeof moduleDependencyGraph === 'object').toEqual(true);
    });

    describe('when access by module id to `ModuleDependencyGraph`', () => {
      it('should returns object that contains property `dependencies` with `Set` value', () => {
        expect(
          moduleDependencyGraph[moduleId].dependencies instanceof Set,
        ).toEqual(true);
      });

      it('should returns object that contains property `inverseDependencies` with `Set` value', () => {
        expect(
          moduleDependencyGraph[moduleId].inverseDependencies instanceof Set,
        ).toEqual(true);
      });

      it('should match snapshot', () => {
        const moduleId = dependencyManager.getModuleId(TEST_MODULE)!;
        expect(moduleDependencyGraph[moduleId]).toMatchSnapshot();
      });
    });
  });

  describe('getInverseDependencies', () => {
    let moduleId: ModuleId;

    beforeEach(() => {
      moduleId = dependencyManager.getModuleId(TEST_MODULE)!;
    });

    it('should match snapshot', () => {
      expect(
        dependencyManager.getInverseDependencies(moduleId),
      ).toMatchSnapshot();
    });
  });
});
