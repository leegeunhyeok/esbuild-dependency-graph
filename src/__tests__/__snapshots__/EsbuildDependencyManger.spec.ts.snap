// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsbuildDependencyManager getDependencyGraph when access by module id to \`ModuleDependencyGraph\` should match snapshot 1`] = `
{
  "dependencies": Set {
    510,
    74,
    449,
    519,
    1043,
    1122,
    1357,
  },
  "inverseDependencies": Set {
    1361,
  },
}
`;

exports[`EsbuildDependencyManager getInverseDependencies should match snapshot 1`] = `
[
  1358,
  1361,
  1362,
  1363,
  1367,
  0,
]
`;
